### ISSUE TITLE
Lint which reports publicly accessible late fields and variables w/o initializers

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#2443]
**Describe the rule you'd like to see implemented**

Several members of the Dart team have noticed that users are declaring `late` and `late final` fields and variables too eagerly, and there are two new error cases that users can run into when using these.

* Accessing a `late` field or variable before it has been initialized is a runtime error.
* Writing a value to a `late final` field or variable after it has been initialized is a runtime error.

There is no mechanism by which the developer of a `late` field or variable can convey to a user of said field or variable whether it has been initialized, or how a user _should_ read from or write to the variable, other than comments. A lint will help developers avoid declaring publicly accessible `late` fields or variables. 

**Examples**

```dart
class C {
  late int a; // LINT
  late final int b; // LINT
  late final int c = 1; // OK?
  late int _d; // OK
  static late int e; // LINT
}
late int f; // LINT
```

I think the lint cases include:

* a public top-level variable declared `late`, without an initializer, not annotated with `@visibleForTesting`, `@visibleForTemplate`, or `@internal`
* a public instance field of a public or private class or mixin declared `late`, without an initializer, not annotated with `@visibleForTesting`, `@visibleForTemplate`, or `@internal`
* a public static field of a public class, mixin, or extension declared `late`, without an initializer, not annotated with `@visibleForTesting`, `@visibleForTemplate`, or `@internal`

**Comment by bwilkerson** [comment]
It would be interesting to know whether there are good uses for this style that would result in false positives, but we probably don't have enough experience yet with the feature to know for sure.

> `late final int c = 1; // OK?`

Yes, I think that should be OK, either with or without the `final`. The initializer is basically a default value that's lazily computed, right?

**Comment by srawlins** [comment]
> The initializer is basically a default value that's lazily computed, right?

Yeah and I think we have static analysis diagnostics for re-assigning such a field. I'm not sure if its in the spec.
